pkg/tsdb/grafana-postgresql-datasource/sqleng/sql_engine_pgx.go

```diff
@@ -0,0 +1,550 @@
+	results []*pgconn.Result, interpolatedQuery string) (*dataQueryModel, error) {
+	columnNames := []string{}
+	columnTypesPGX := []string{}
+
+	// The results will contain column information in the metadata
+	for _, result := range results {
+		// Get column names from the result metadata
+		for _, field := range result.FieldDescriptions {
+			columnNames = append(columnNames, field.Name)
+			pqtype, ok := pgtype.NewMap().TypeForOID(field.DataTypeOID)
+			if !ok {
+				// Handle special cases for field types
+				if field.DataTypeOID == pgtype.TimetzOID {
+					columnTypesPGX = append(columnTypesPGX, "timetz")
+				} else if field.DataTypeOID == 790 {
+					columnTypesPGX = append(columnTypesPGX, "money")
+				} else {
+					return nil, fmt.Errorf("unknown data type oid: %d", field.DataTypeOID)
+				}
+			} else {
+				columnTypesPGX = append(columnTypesPGX, pqtype.Name)
+			}
+		}
+	}
+
+	qm := &dataQueryModel{
+		columnTypesPGX: columnTypesPGX,
+		columnNames:    columnNames,
+		timeIndex:      -1,
+		timeEndIndex:   -1,
+		metricIndex:    -1,
+		metricPrefix:   false,
+		queryContext:   queryContext,
+	}
+
+	queryJson := QueryJson{}
+	err := json.Unmarshal(query.JSON, &queryJson)
+	if err != nil {
+		return nil, err
+	}
+
+	if queryJson.Fill {
+		qm.FillMissing = &data.FillMissing{}
+		qm.Interval = time.Duration(queryJson.FillInterval * float64(time.Second))
+		switch strings.ToLower(queryJson.FillMode) {
+		case "null":
+			qm.FillMissing.Mode = data.FillModeNull
+		case "previous":
+			qm.FillMissing.Mode = data.FillModePrevious
+		case "value":
+			qm.FillMissing.Mode = data.FillModeValue
+			qm.FillMissing.Value = queryJson.FillValue
+		default:
+		}
+	}
+
+	qm.TimeRange.From = query.TimeRange.From.UTC()
+	qm.TimeRange.To = query.TimeRange.To.UTC()
+
+	switch queryJson.Format {
+	case "time_series":
+		qm.Format = dataQueryFormatSeries
+	case "table":
+		qm.Format = dataQueryFormatTable
+	default:
+		panic(fmt.Sprintf("Unrecognized query model format: %q", queryJson.Format))
+	}
+
+	for i, col := range qm.columnNames {
+		for _, tc := range e.timeColumnNames {
+			if col == tc {
+				qm.timeIndex = i
+				break
+			}
+		}
+
+		if qm.Format == dataQueryFormatTable && strings.EqualFold(col, "timeend") {
+			qm.timeEndIndex = i
+			continue
+		}
+
+		switch col {
+		case "metric":
+			qm.metricIndex = i
+		default:
+			if qm.metricIndex == -1 {
+				columnType := qm.columnTypesPGX[i]
+				for _, mct := range e.metricColumnTypes {
+					if columnType == mct {
+						qm.metricIndex = i
+						continue
+					}
+				}
+			}
+		}
+	}
+	qm.InterpolatedQuery = interpolatedQuery
+	return qm, nil
+}
+
+func convertResultsToFrame(results []*pgconn.Result, query string, rowLimit int64) (*data.Frame, error) {
+	frame := data.Frame{}
+	m := pgtype.NewMap()
+
+	for _, result := range results {
+		// Skip non-select statements
+		if !result.CommandTag.Select() {
+			continue
+		}
+		fieldDescriptions := result.FieldDescriptions
```

Comment:
nit: Do we need to create `fieldDescriptions`? Can we just use `result.FieldDescriptions`?

Reply:
using `result.FieldDescriptions` should be fine.